Sprint 8 solution time and duration

IZarubitskiy opened the pull request:
ТЗ-8 PullRequest

IZarubitskiy committed:
ТЗ-8

Начало реализации. осталось еще много ошибок связанных с добавлением задач

IZarubitskiy committed:
ТЗ-8

Начата реализация времени - промежуточное сохранение.

IZarubitskiy committed:
ТЗ-8

Начата реализация времени - промежуточное сохранение.

IZarubitskiy committed:
ТЗ-8

Начата реализация времени - промежуточное сохранение.

IZarubitskiy committed:
ТЗ-8

Начата реализация времени - промежуточное сохранение.

IZarubitskiy committed:
4 вариант(2)
Исправил конструктор.
Поднял выше метод загрузки из файла

Теперь принимает файл и работает уже только с ним

IZarubitskiy committed:
ТЗ-8 Первая итерация

IZarubitskiy committed:
ТЗ-8 Первая итерация

IZarubitskiy committed:
ТЗ-8 Первая итерация

IZarubitskiy committed:
ТЗ-8 Первая итерация

IZarubitskiy committed:
ТЗ-8 Первая итерация

IZarubitskiy committed:
ТЗ-8 Первая итерация

PolnySkvorets commented on test/InMemoryTaskManagerTest.java:
```diff
@@ -6,10 +6,10 @@
     InMemoryTaskManager inMemoryTaskManager = new InMemoryTaskManager();
-
+/*
     @Test
```
Нужно починить тест, если он должен работать

PolnySkvorets commented on test/FileBackedTaskManagerTest.java:
```diff
@@ -21,20 +24,59 @@ class FileBackedTaskManagerTest {
     @Test
     void save() {
-        System.out.println("Пустая строка перед файлом");
+        //System.out.println("Пустая строка перед файлом");
```
Нужно почистить от дебага

PolnySkvorets commented on test/FileBackedTaskManagerTest.java:
```diff
@@ -21,20 +24,59 @@ class FileBackedTaskManagerTest {
+        System.out.println("задача 5");
+        fileBackedTaskManager.addTask(task5);
+        System.out.println("задача 6");
+        fileBackedTaskManager.addTask(task6);
```
Не нужно дебаг вывод оставлять в тестах

PolnySkvorets commented on test/InMemoryTaskManagerTest.java:
```diff
@@ -6,10 +6,10 @@
 class InMemoryTaskManagerTest {
```
Не хвататет по требованиям покрытия тестами

PolnySkvorets reviewed (commented):
Оставил несколько замечаний

IZarubitskiy committed:
ТЗ-8 Работа над ошибками (1)

IZarubitskiy committed:
ТЗ-8:

Работа над ошибками, покрытие тестами

IZarubitskiy committed:
ТЗ-8:

Работа над ошибками, покрытие тестами

IZarubitskiy committed:
ТЗ-8:

Работа над ошибками, покрытие тестами

IZarubitskiy committed:
ТЗ-8:

Работа над ошибками, покрытие тестами

IZarubitskiy committed:
ТЗ-8:

Работа над ошибками, покрытие тестами

IZarubitskiy committed:
ТЗ-8:

Работа над ошибками, покрытие тестами

IZarubitskiy committed:
ТЗ-8:

Работа над ошибками, покрытие тестами

IZarubitskiy committed:
ТЗ-8:

Работа над ошибками, покрытие тестами

IZarubitskiy committed:
ТЗ-8:

Работа над ошибками, покрытие тестами

IZarubitskiy committed:
ТЗ-8:

Работа над ошибками, покрытие тестами

IZarubitskiy committed:
ТЗ-8:

Работа над ошибками, покрытие тестами

IZarubitskiy committed:
ТЗ-8:

Работа над ошибками, покрытие тестами

IZarubitskiy committed:
ТЗ-8 Работа над ошибками (1)

IZarubitskiy committed:
ТЗ-8 Работа над ошибками (1)

IZarubitskiy committed:
ТЗ-8 Работа над ошибками (1)

IZarubitskiy committed:
ТЗ-8 Работа над ошибками (1)

IZarubitskiy committed:
ТЗ-8 Работа над ошибками (1)

IZarubitskiy committed:
ТЗ-8 Работа над ошибками (1)

IZarubitskiy committed:
ТЗ-8 Работа над ошибками (1)

IZarubitskiy committed:
ТЗ-8 Работа над ошибками (1)

IZarubitskiy committed:
ТЗ-8 Работа над ошибками (1)

IZarubitskiy committed:
ТЗ-8 Работа над ошибками (1)

IZarubitskiy committed:
ТЗ-8 Работа над ошибками (1)

IZarubitskiy committed:
ТЗ-8 Работа над ошибками (1)

IZarubitskiy committed:
ТЗ-8 Работа над ошибками (1)

IZarubitskiy committed:
ТЗ-8 Работа над ошибками (1)

IZarubitskiy committed:
ТЗ-8 Работа над ошибками (1)

IZarubitskiy committed:
ТЗ-8 Работа над ошибками (1)

IZarubitskiy committed:
ТЗ-8 Работа над ошибками (1)

IZarubitskiy committed:
ТЗ-8 Работа над ошибками (1)

PolnySkvorets commented on src/InMemoryHistoryManager.java:
```diff
@@ -31,14 +31,14 @@ public Node linkLast(Task task) {
 
-    public List<Task> getTasks() {
-        List<Task> taskListHystory = new ArrayList<>();
+    public List<Task> getHistory() {
```
Добавь @override

PolnySkvorets commented on src/InMemoryHistoryManager.java:
```diff
@@ -88,10 +88,4 @@ public void remove(int id) {
     }
-
-    @Override
-    public List<Task> getHistory() {
-        history = getTasks();
-        return history;
-    }
```
здесь можно было просто return getTasks(); сделать

IZarubitskiy replied:
странно. я в коммитах нынешних это уже поменял все. почему отображается в гите старое. не понимаю ..

PolnySkvorets commented on src/Epic.java:
```diff
@@ -1,19 +1,23 @@
 
 public class Epic extends Task {
-    ArrayList<Integer> idSubtasklist;
+    ArrayList<Integer> subTaskListId;
+    LocalDateTime endTime;
```
Добавь модификаторы доступа

PolnySkvorets reviewed (commented):
Оставил замечания + напишу в практикуме

IZarubitskiy committed:
ТЗ-8:

Работа над ошибками:
Переместил класы по папкам
Обновил коментариии из GIT

IZarubitskiy committed:
ТЗ-8:

Работа над ошибками:
Переместил клаcсы по папкам
Обновил коментариии из GIT
Добавил тесты для History Manager

IZarubitskiy committed:
ТЗ-8:

Работа над ошибками:
Переместил клаcсы по папкам
Обновил коментариии из GIT
Добавил тесты для History Manager

IZarubitskiy committed:
ТЗ-8:

Работа над ошибками:
Переместил клаcсы по папкам
Обновил коментариии из GIT
Добавил тесты для History Manager

PolnySkvorets commented on test/managers/TaskManagerTest.java:
```diff
@@ -0,0 +1,48 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotEquals;
+
+abstract class TaskManagerTest<T extends TaskManager>{
```
По идее здесь ожидается, что будут реализованы тестовые методы. То есть нужно добавить например абстрактный  метод getManager, который будет возвращать интерефейс Manager

Manager getManager();

добавить его в реализацию дочерних тестов, например для InMemoryTaskManager

@Override
Manager getManager(){
return new InMemoryTaskManager();
}

PolnySkvorets replied:
Да и  FileBackedTaskManagerTest тоже должен наследоваться от этого теста

PolnySkvorets commented on test/managers/TaskManagerTest.java:
```diff
@@ -0,0 +1,48 @@
+    void historyTests() {}
+
+    @Test
+    void addNewTask() {}
```
А здесь в этих методах уже реализация:
@Test
    void addNewTask() {
Manager manager = getManager();
manager.addTask();
.....
assert(.....
}

PolnySkvorets replied:
Поэтому по требованиям задачи это должен быть абстрактный класс, в котором реализованы повторяющиеся тесты для InMemoryTaskManager и FileBackedTaskManager.

PolnySkvorets commented on test/managers/TaskManagerTest.java:
```diff
@@ -0,0 +1,48 @@
+
+import org.junit.jupiter.api.Test;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotEquals;
+
```
Проверяй и не оставляй неиспользуемые импорты

IZarubitskiy replied:
я что-то расслабился что гит обычно на это указывает . и не проверил

PolnySkvorets commented on test/managers/InMemoryHistoryManagerTest.java:
```diff
@@ -0,0 +1,66 @@
+                TaskStatus.NEW, LocalDateTime.parse("01.02.2031, 13:50", DateTimeFormatter.ofPattern("dd.MM.yyyy, HH:mm")), Duration.ofMinutes(60)));
+
+        inMemoryHistoryManager.add(new Task("Test addNewTask 2", "Test addNewTask 2 description", 30,
+                TaskStatus.NEW, LocalDateTime.parse("01.02.2032, 13:50", DateTimeFormatter.ofPattern("dd.MM.yyyy, HH:mm")), Duration.ofMinutes(60)));
+        inMemoryHistoryManager.add(new Task("Test addNewTask 3", "Test addNewTask 3 description", 31,
+                TaskStatus.NEW, LocalDateTime.parse("01.02.2033, 13:50", DateTimeFormatter.ofPattern("dd.MM.yyyy, HH:mm")), Duration.ofMinutes(60)));
+        inMemoryHistoryManager.add(new Task("Test addNewTask 1", "Test addNewTask 1 description", 29,
+                TaskStatus.NEW, LocalDateTime.parse("01.02.2031, 13:50", DateTimeFormatter.ofPattern("dd.MM.yyyy, HH:mm")), Duration.ofMinutes(60)));
```
Здесь не обязательно добавлять недублирующиеся задачи, достаточно двараза одну и ту же задачу добавить, чтобы проверить.

PolnySkvorets commented on test/managers/InMemoryHistoryManagerTest.java:
```diff
@@ -0,0 +1,66 @@
+        assertNotEquals(34, inMemoryHistoryManager.getHistory().getLast().getId(), "задача не удалена с начала");
+
+        inMemoryHistoryManager.remove(32);
+        assertEquals(2, inMemoryHistoryManager.getHistory().size(), "Задача удалена");
+        assertEquals(33, inMemoryHistoryManager.getHistory().getFirst().getId(), "Удалена не та задача");
+        assertEquals(31, inMemoryHistoryManager.getHistory().getLast().getId(), "задача не удалена с конца");
```
Лучше проверить просто получением задачи по id, что после удаления она не найдена

IZarubitskiy replied:
тут так сделать не получится - HistoryManager по заданию создавался на нодах - мы делали двусвязный список и потому только по таким образом могу проверить  работу HistoryManager

PolnySkvorets reviewed (commented):
Оставил замечания

IZarubitskiy committed:
ТЗ-8:

Создание абстрактного класса

IZarubitskiy committed:
ТЗ-8 Работа над ошибками (1)

IZarubitskiy committed:
ТЗ-8 Работа над ошибками (1)

PolnySkvorets reviewed (approved):
Принято
